sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java

```diff
@@ -531,6 +532,30 @@ public void process(
                       element.getKey().getKey(), dynamicDestinations, datasetService);
               tableSchema = converter.getTableSchema();
               descriptor = converter.getDescriptor(false);
+
+              if (autoUpdateSchema) {
```

Comment:
I'm not sure this is the ideal place to put this. getAppendClientInfo is called whenever the static cache is populated, meaning that on any worker restart, range move, etc. we'll be forced to call this API again. However we have persistent state in this DoFn, so we know if it's a "new" key or not. Can we use that to gate calling this method instead?

Reply:
We should always perform this check before creating a new StreamWriter , regardless of the reason for its creation. The only exception is if we already have an updated schema stored in state (see first `if` block above). If I'm following correctly, this method (`getAppendClientInfo`) will always [create a new stream writer](https://github.com/apache/beam/blob/d17f77a015d28b3008e80e647cf5e230310333c7/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/StorageApiWritesShardedRecords.java#L535-L554).

Reply:
Also note that the updated schema is ignored when the **StreamWriter object's** creation time is later than the updated schema's. 
i.e. it doesn't matter when the WriteStream itself was created

Reply:
One optimization we can do is store a destination's schema in cache so that it's accessible to multiple StreamWriters writing to the same table on the same machine